### DISCUSSION TITLE
Chi square tests following multiple imputation

### DISCUSSION THREAD

**Discussion by franzepsy** [amices/mice#620] [Impute, analyse and pool]
I have conducted multiple imputation with 'mice' and now need to conduct chi square tests (factor with 2 levels x factor with 4 levels) in each of the (32) multiply imputed datasets and pool the results. 

I am aware of the formula:
dk <- c(24.957, 18.051, 18.812, 17.362, 21.234, 18.615, 19.84)
dk.comb <- miceadds::micombine.chisquare(dk=dk, df=4 )

But how can I obtain the chi square results for each of the 32 imputed datasets, to then pool them together? In short, is there a formula to obtain dk?

Thank you!

**Comment by hanneoberman** [comment]
If you want to obtain estimates on each imputed dataset, you can use functional programming (or a for loop, if you're more comfortable with that).

Here a small reprex:

``` r
# setup
library(mice)
#> Warning: package 'mice' was built under R version 4.3.1
#> 
#> Attaching package: 'mice'
#> The following object is masked from 'package:stats':
#> 
#>     filter
#> The following objects are masked from 'package:base':
#> 
#>     cbind, rbind
# incomplete data
dat <- nhanes
dat$age <- as.factor(dat$age)
dat$hyp <- as.factor(dat$hyp)
chisq.test(x = dat$age, y = dat$hyp)
#> Warning in chisq.test(x = dat$age, y = dat$hyp): Chi-squared approximation may
#> be incorrect
#> 
#>  Pearson's Chi-squared test
#> 
#> data:  dat$age and dat$hyp
#> X-squared = 4.7731, df = 2, p-value = 0.09195
# imputed data
imp <- mice(dat, print = FALSE)
compl <- complete(imp, "all")
purrr::map(compl, ~{
  chisq.test(x = .x$age, y = .x$hyp)
})
#> Warning in chisq.test(x = .x$age, y = .x$hyp): Chi-squared approximation may be
#> incorrect
#> Warning in chisq.test(x = .x$age, y = .x$hyp): Chi-squared approximation may be
#> incorrect

#> Warning in chisq.test(x = .x$age, y = .x$hyp): Chi-squared approximation may be
#> incorrect

#> Warning in chisq.test(x = .x$age, y = .x$hyp): Chi-squared approximation may be
#> incorrect

#> Warning in chisq.test(x = .x$age, y = .x$hyp): Chi-squared approximation may be
#> incorrect
#> $`1`
#> 
#>  Pearson's Chi-squared test
#> 
#> data:  .x$age and .x$hyp
#> X-squared = 8.2916, df = 2, p-value = 0.01583
#> 
#> 
#> $`2`
#> 
#>  Pearson's Chi-squared test
#> 
#> data:  .x$age and .x$hyp
#> X-squared = 9.0561, df = 2, p-value = 0.0108
#> 
#> 
#> $`3`
#> 
#>  Pearson's Chi-squared test
#> 
#> data:  .x$age and .x$hyp
#> X-squared = 5.9524, df = 2, p-value = 0.05099
#> 
#> 
#> $`4`
#> 
#>  Pearson's Chi-squared test
#> 
#> data:  .x$age and .x$hyp
#> X-squared = 3.3351, df = 2, p-value = 0.1887
#> 
#> 
#> $`5`
#> 
#>  Pearson's Chi-squared test
#> 
#> data:  .x$age and .x$hyp
#> X-squared = 9.8828, df = 2, p-value = 0.007144
```

<sup>Created on 2024-02-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Reply by franzepsy** [reply]
Thank you, that is helpful. But is there a way to automatically extract the chi square results from each imputed dataset, to then pool them together? The process for my analyses would be really long, considering that I would get 32 values for each of the 20 chi square tests that I need to run, so I was wondering if there is an automatic way to do it, instead of manually extracting the values. Thank you again.

**Comment by hanneoberman** [comment]
There is no *automatic* way to do this, but a small edit to the above example solves the pooling step:

```{r}
ests <- purrr::map_dbl(compl, ~{
  chisq.test(x = .x$age, y = .x$hyp)$statistic
})
pooled <- miceadds::micombine.chisquare(ests, df = 2)

#> Combination of Chi Square Statistics for Multiply Imputed Data
#> Using 5 Imputed Data Sets
#> F(2, 81.29)=2.794     p=0.06709 
```

Then you can just map or loop over the 20 variables you're describing.

**Reply by franzepsy** [reply]
Thank you. That works when I try with your example, but I get an error message when I use my data. Could it be because my 'x' factor has 4 levels, and my 'y' factor has 2 levels?

This is the message that I get when I run the function purrr::map_dbl() : 
> rlang::last_trace()
<error/purrr_error_indexed>
Error in `purrr::map_dbl()`:
ℹ In index: 2.
ℹ With name: imp.
Caused by error in `chisq.test()`:
! at least one entry of 'x' must be positive
---
Backtrace:
    ▆
 1. └─purrr::map_dbl(...)
 2.   └─purrr:::map_("double", .x, .f, ..., .progress = .progress)
 3.     ├─purrr:::with_indexed_errors(...)
 4.     │ └─base::withCallingHandlers(...)
 5.     ├─purrr:::call_with_cleanup(...)
 6.     └─global .f(.x[[i]], ...)
 7.       └─stats::chisq.test(x = .x$Group_num, y = .x$f_highest_ed_qualification_18wgest_1degree2other)
 8.         └─base::stop("at least one entry of 'x' must be positive")
Run rlang::last_trace(drop = FALSE) to see 4 hidden frames.
